src/libp2p/connection/yamux.rs

```diff
@@ -895,6 +904,48 @@ pub struct Config {
     pub randomness_seed: [u8; 16],
 }
 
+/// Reference to a substream within the [`Yamux`].
+// TODO: Debug
+pub struct SubstreamRef<'a, T> {
+    id: SubstreamId,
+    substream: &'a Substream<T>,
+}
+
+impl<'a, T> SubstreamRef<'a, T> {
+    /// Identifier of the substream.
+    pub fn id(&self) -> SubstreamId {
+        self.id
+    }
+
+    /// Returns the user data associated to this substream.
+    pub fn user_data(&self) -> &T {
+        &self.substream.user_data
+    }
+
+    /// Returns the user data associated to this substream.
+    pub fn into_user_data(self) -> &'a T {
```

Comment:
_don't see much difference between this and above fn except `'a` lifetime._

Reply:
Yeah but they're fundamentally different.

These are methods on an object that borrows the connection.
`user_data(&self) -> &T` returns a `&T` that borrows the object that borrows the connection, while `into_user_data(self) -> &'a T` kills the object and returns a `&T` that borrows the connection.

Reply:
Oh, right. Thanks